### DISCUSSION TITLE
Why is the initial migration scanning ALL my schemas and taking 5-10min to complete?

### DISCUSSION THREAD

**Discussion by Frank-D** [mikro-orm/mikro-orm#3021] [Q&A] [answered]
Hi,

I've started my backend with node/nestjs/graphql & mikro-orm (5.1), and currently I'm only having one 'User' entity with crud operations. I'm currently only working on all the plumbing, and got it all working.

However, as I am adding new fields to my user entity as my app/logic evolves over time, instead of generating additional incremental migration files, since I still haven't released anything yet to production (and far from it!), I would rather have one single clean initial migration file for now, so I keep deleting my existing migration files (first migration file and the snapshot) and cleaning the database using a little sql script.

That process works well in theory, however in practice, every time I execute the mikro-orm initial migration again (e.g. `npx mikro-orm migration:create --initial`), it seems like instead of ONLY checking against my postgres (cockroachdb more precisely..) "public" empty application schema, which I've provided/set btw as part of my mikro-orm options, the migration process is scanning ALL the accessible metadata schemas under my provided « dbName » (fyi, every cochroachdb database created will come with 1 application empty « public » schema, plus 3 or 4 metadata schemas full of tables….I believe this is very similar to how postgres works too..).

I can see that happening by using a database monitoring tool, and seeing all those queries being done by the migration process against all those other schemas which have nothing to do with the application I'm building, and they are not linked whatsoever inside my application nor in any entities or mikro-orm configuration. The migration logic seems to be looping over many metadata tables of other irrelevant schemas, even my cpu starts running high at some point.

And that is the reason why such initial migration process is taking precisely 8 minutes(!!) to complete, and in the end, only to produce one single ~DDL statement, which is, the creation of my single 'Users' table! (against the right ‘public’ schema….so the migration output is ok/accurate, in the end of all this!..).

I'm trying to understand why this is happening under the hood, I checked the underlying mikro-orm migration source code a bit, but since it's talking under the hood to knex and postgres driver and other libs, it's hard to follow and understand why this is happening.

Any help in preventing all those extra schemas scanning (which brings absolutely no value but just slowling down big time the initial migration process..) would be super appreciated. 

Thank you!

**Comment by Frank-D** [comment, marked as answer]
Providing more info:

> ## My dependencies

Dependency | Version
-- | --
node | 16.13.1
typescript | 4.5.5
mikro-orm | 5.1.2
@mikro-orm/postgresql | 5.1.2


> ## My ORM config:
```JavaScript
const config: Options = {
    type: 'postgresql',
    host: 'localhost',
    port: 26257, 
    dbName: 'myappdb',
    schema: 'public',
    schemaGenerator: { disableForeignKeys: false },
    user: 'myappuser',
    // password: process.env.DBPASSWORD || 'postgres',
    entities: ['./dist/**/*.entity.js'],
    entitiesTs: ['./src/**/*.entity.ts'],
    metadataProvider: TsMorphMetadataProvider,
    debug: true,
    logger: logger.log.bind(logger),
    highlighter: new SqlHighlighter(),
    driverOptions: { version: 13 }, // required by knex under the hood, or else it will assume db is really 'postgresql' instead of "~a version of postgresql such as cockroachdb"...
    migrations: {
        tableName: 'mikro_orm_migrations', // name of database table with log of executed transactions
        path: './dist/resources/migrations', // path to the folder with .js compiled migrations
        pathTs: './resources/migrations', // path to the folder with .ts migrations
        //pattern: /^[\w-]+\d+\.ts$/, // regex pattern for the migration files --> v4
        //glob: '!(*.d).{js,ts}', // how to match migration files (all .js and .ts files, but not .d.ts) --> v5
        transactional: true, // wrap each migration in a transaction
        disableForeignKeys: false, // wrap statements with `set foreign_key_checks = 0` or equivalent; for postgresql, this generates "set session_replication_role = 'replica';" migration sql statements, which fail when ran against cockroachdb! So need to turn off this setting..
        allOrNothing: true, // wrap all migrations in master transaction
        dropTables: true, // allow to disable table dropping
        safe: true, // allow to disable table and column dropping
        emit: 'ts', // migration generation mode
        generator: TSMigrationGenerator,
    },
};
```

> ## My DB
- Cockroachdb v21.2.3
- Default schemas after initial manual db installation:

OOTB Schemas | Purpose
-- | --
crdb_internal | System
information_schema | System
pg_catalog | System
pg_extension | System
public | Application

> ## My App Entity (stripped down version)
```TypeScript
import { Entity, PrimaryKey, Property, Enum, Unique } from '@mikro-orm/core';
import { v4 as UUIDv4 } from 'uuid';
import { CustomUserRepository } from './user.repository';

@Entity({
    tableName: 'users',
    schema: 'public',
    customRepository: () => CustomUserRepository,
})
export class UserEntity {
    constructor(
        authId: string,
        email: string,
        fullName: string,
    ) {
        this.authId = authId;
        this.email = email;
        this.fullName = fullName;
    }

    @PrimaryKey()
    id = UUIDv4();

    @Property()
    createdOn = new Date();

    @Property({ onUpdate: () => new Date() })
    updatedOn = new Date();

    @Property()
    lastLogin = new Date();

    @Property()
    @Unique()
    authId: string;

    @Property()
    @Unique()
    email!: string;

    @Property()
    fullName: string;

    @Property({ onUpdate: (user: UserEntity) => user.version + 1 })
    version = 1;
}
```

> ## Generated 'create --initial' Migration DDL file
```TypeScript
import { Migration } from '@mikro-orm/migrations';

export class Migration20220416155921 extends Migration {

  async up(): Promise<void> {
    this.addSql('create table "users" ("id" varchar(255) not null, "created_on" timestamptz(0) not null, "updated_on" timestamptz(0) not null, "last_login" timestamptz(0) not null, "auth_id" varchar(255) not null, "email" varchar(255) not null, "full_name" varchar(255) not null, "version" int not null);');
    this.addSql('alter table "users" add constraint "users_auth_id_unique" unique ("auth_id");');
    this.addSql('alter table "users" add constraint "users_email_unique" unique ("email");');
    this.addSql('alter table "users" add constraint "users_pkey" primary key ("id");');
  }

}
```

> ## How to reproduce
Simply launching the mikro-orm migration > create > initial command from the command line:
```Shell
npx mikro-orm migration:create --initial
```
- empty cockroach db (..which contains the default system schemas etc, see my initial msg for more info.)
- only one single entity in my app, expected to generate one single ~ 'create table' ddl statement

> ### Expected

1. One simple ddl migration file created, containing one single ~'create table' statement for my one single app entity
2. Migration logic to scan 'only' my 'public' cochroachdb schema which I've specified in my ORM config (which is my app schema, and totally empty / no tables / fresh start..)
3. Migration execution time: anywhere between ~2-20 seconds

> ### Actual
1. `[ OK ]` One simple ddl migration file created, containing one single ~'create table' statement for my one single app entity
2. `[ NOT OK ]` Migration logic ignores my specified 'public' schema from my ORM config, and scans ALL my cochroachdb schemas including all db system schemas containing many many internal tables which have nothing to do with my app, slowing down the migration execution a lot..
3. `[ NOT OK ]` Migration execution time: anywhere between `~7-8 **minutes**`

**Reply by Frank-D** [reply]
Hello @B4nan, 
I know you are super busy and all over the place in this project/repo, so I waited a month before pinging you directly on this (was trying to avoid pinging you..:/ ). 

I've added a new comment right above following my original post, and tried to provide as much details as possible, but please let me know if I should add anything else to help you or anyone else help me with that odd perf. issue. Any help much appreciated, but of course no pressure!;-)

thanks

---
Some different but +- related issue/discussion:
- https://github.com/mikro-orm/mikro-orm/issues/386#issuecomment-1121065610
- https://github.com/mikro-orm/mikro-orm/discussions/3102

**Reply by B4nan** [reply]
I can't help much with cockroach, its not officially supported and given they dont provide ARM images, I can't even verify things locally as the existing docker images that work on M1 are extremely slow (tried even some community ones, still unusable). Once there will be usable docker image for M1 macs, I am more than happy to provide official driver (given knex now ofifically supports it too), already have a WIP for it but the slow docker image makes it unmergable.

Few notes:
- schema diffing _should_ go thru all your schemas, this is correct behaviour in postgres, as we support entities in multiple schemas
- the `schema` option you specify in your config is only the default one, that does not mean "use only one schema and ignore rest"
- if you use latest dev version, just yesterday I added an option to list schemas that should be ignored via `schemaGenerator.ignoreSchema: [...]`
- I am willing to ignore schemas prefixed with `crdb_` just like we are already ignoring those prefixed with `pg_`, I guess we can do that right now to ease things instead of waiting for cockroach native driver (edit: [here you go](https://github.com/mikro-orm/mikro-orm/commit/049fea3f77063f504a617e32cb9a4e303fbbf0be))

> Migration execution time: anywhere between ~7-8 **minutes**

Maybe you are also on M1? :] Slowness of cockroach is definitely not related to the ORM code - migrations are just about executing the queries you have there.

**Reply by Frank-D** [reply]
Thank you @B4nan for your great response! (and I have to also add, thank you so much for this amazing ORM!)

> I can't help much with cockroach, its not officially supported and given they dont provide ARM images, I can't even verify things locally as the existing docker images that work on M1 are extremely slow (tried even some community ones, still unusable).

I see and understand your point/pain.
Not sure if you tried this below specific cockroachdb ARM image from the community, they seem to be pretty excited about it, so I'm assuming it's working fine but who knows...
--> https://github.com/cockroachdb/cockroach/issues/68648#issuecomment-992225541

> Once there will be usable docker image for M1 macs, I am more than happy to provide official driver (given knex now ofifically supports it too), already have a WIP for it but the slow docker image makes it unmergable.

Yes exactly since knex now officially supports cockroach, it would be amazing to eventually see mikro officially supporting it as well. 
At least I'm glad to see that you're open to eventually add support for it, good to know..

> schema diffing should go thru all your schemas, this is correct behaviour in postgres, as we support entities in multiple schemas
> the schema option you specify in your config is only the default one, that does not mean "use only one schema and ignore rest"

Ok, makes sense.

> if you use latest dev version, just yesterday I added an option to list schemas that should be ignored via schemaGenerator.ignoreSchema: [...]

Amazing, that's what I actually need for now, I'll definitely give it a try.

> I am willing to ignore schemas prefixed with crdb_ just like we are already ignoring those prefixed with pg_, I guess we can do that right now to ease things instead of waiting for cockroach native driver (edit: [here you go](https://github.com/mikro-orm/mikro-orm/commit/049fea3f77063f504a617e32cb9a4e303fbbf0be))

Even much much better!! :D 🙏 🙌

> Maybe you are also on M1? :] Slowness of cockroach is definitely not related to the ORM code - migrations are just about executing the queries you have there.

Nope I'm on standard intel i7 (mac pro)...but I've used the web-portal cockroachdb traffic analytic tool that comes ootb with cockroach (and which is very convenient and easy to use btw..), and this is how I could identified the hundreds and hundreds of queries being executed against all those 'system' schemas when I trigger that migration:create --initial command. If useful to you, I could even record a short screencast of that web db traffic analytic tool in action as soon as I trigger such mikro migration command, but I guess at this point, it's pointless, with the previous response of the 2 quotes above, that should be more than enough to fix this situation! ; )

Thanks again!!

**Reply by B4nan** [reply]
> Not sure if you tried this below specific cockroachdb ARM image from the community, they seem to be pretty excited about it, so I'm assuming it's working fine but who knows...

Yes, I tried everything from that issue. It's all still extremely slow, and there were no updates to those images since then, so I don't think anything changed. 

If you are on intel, it feels weird that execution of migrations would be so slow, but I really have no experience with cockroach so not sure what performance to expect in the first place. Have you tried to execute those 4 queries in some 3rd party tool?

**Reply by Frank-D** [reply]
I haven’t tried to sort of “record” all those queries being triggered by the mikro migration process so that I can then manually try to execute them myself no, but to me, the performance issue totally seemed to be related to the very large quantity of db queries being executed against all cochroach system schemas (much more than “4 queries”, we’re talking here about many hundreds, which makes it last for 7-8 mins..), rather than performance issue related to any specific query in particular..

On mobile right now going to the office for the day, but I’ll try later to provide some more insignts about the number of tables in each of those schemas, and perhaps more useful, I’ll see if I can record all the queries being made while triggering the migration process, if this can provide any help… 

But ultimately, I see no good reason why someone would want the mikro migration process to scan/diff their cochroach ‘system’ schemas full of internal cochroach tables, no link with the application someone would be trying to build, so I still believe the real effective solution here, is to simply skip scanning those internal cochraoch system schemas, which you already provided a fix/commit above, and I thank you for that.

**Reply by B4nan** [reply]
> very large quantity of db queries being executed against all cochroach system schemas (much more than “4 queries”, we’re talking here about many hundreds, which makes it last for 7-8 mins..), rather than performance issue related to any specific query in particular..

But you said executing migrations is slow too, and executing migrations does not check your current schema at all, it should really translate to those 4 queries + few more checking the migrations table state and ensuring that it already exists.

Btw the `crdb_` prefix is alredy skipped in latest dev version, so just try that and see if it helps before you measure anything.

**Reply by Frank-D** [reply]
hi @B4nan,
this is just to report back that I've updated to 5.1.4, and your fix is working perfectly for cockroach, e.g. my `npx mikro-orm migration:create --initial` command went down, from 7-8 minutes to.......exactly 10 seconds! ; )  ..and validated the output, it produced the exact same migration script as before, which is perfect. It just no longer scans all cockroach internal/system schemas/tables, and that's amazing. Thank you very much one more time, cheers.
